src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -183,230 +138,172 @@ uct_cuda_ipc_get_remote_cache(pid_t pid, uct_cuda_ipc_cache_t **cache)
-                      cache->name, UCS_PGT_REGION_ARG(&region->super),
-                      (void *)key->d_bptr, key->b_len);
-
-            status = ucs_pgtable_remove(&cache->pgtable, &region->super);
-            if (status != UCS_OK) {
-                ucs_error("%s: failed to remove address:%p from cache",
-                          cache->name, (void *)key->d_bptr);
-                goto err;
-            }
-
-            /* close memhandle */
-            UCT_CUDADRV_FUNC_LOG_ERR(
-                    cuIpcCloseMemHandle((CUdeviceptr)region->mapped_addr));
-            ucs_free(region);
+UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_ipc_map_memhandle,
+                 (md, key, mapped_addr, region_p),
+                 uct_cuda_ipc_md_t *md,
+                 const uct_cuda_ipc_key_t *key, void **mapped_addr,
+                 uct_cuda_ipc_rcache_region_t **region_p)
+{
+    ucs_rcache_t *cache;
+    ucs_status_t status;
+
+    if (md->rcache_enable != UCS_NO) {
+        status = uct_cuda_ipc_get_remote_cache(md, key->pid, &cache);
+        if (status != UCS_OK) {
+            return status;
         }
-    }
 
-    status = uct_cuda_ipc_open_memhandle(key, (CUdeviceptr*)mapped_addr);
-    if (ucs_unlikely(status != UCS_OK)) {
-        if (ucs_likely(status == UCS_ERR_ALREADY_EXISTS)) {
-            /* unmap all overlapping regions and retry*/
-            uct_cuda_ipc_cache_invalidate_regions(cache, (void *)key->d_bptr,
-                                                  UCS_PTR_BYTE_OFFSET(key->d_bptr,
-                                                                      key->b_len));
-            status = uct_cuda_ipc_open_memhandle(key,
-                                                 (CUdeviceptr*)mapped_addr);
-            if (ucs_unlikely(status != UCS_OK)) {
-                if (ucs_likely(status == UCS_ERR_ALREADY_EXISTS)) {
-                    /* unmap all cache entries and retry */
-                    uct_cuda_ipc_cache_purge(cache);
-                    status =
-                        uct_cuda_ipc_open_memhandle(key,
-                                                    (CUdeviceptr*)mapped_addr);
-                    if (status != UCS_OK) {
-                        ucs_fatal("%s: failed to open ipc mem handle. addr:%p "
-                                  "len:%lu (%s)", cache->name,
-                                  (void *)key->d_bptr, key->b_len,
-                                  ucs_status_string(status));
-                    }
-                } else {
-                    ucs_fatal("%s: failed to open ipc mem handle. addr:%p len:%lu",
-                              cache->name, (void *)key->d_bptr, key->b_len);
-                }
-            }
-        } else {
-            ucs_debug("%s: failed to open ipc mem handle. addr:%p len:%lu",
-                      cache->name, (void *)key->d_bptr, key->b_len);
-            goto err;
+        status = uct_cuda_ipc_check_rcache(cache, (uct_cuda_ipc_key_t *)key, region_p);
```

Comment:
why need to cast key?

Reply:
Because `key` is of type `const uct_cuda_ipc_key_t *` but `uct_cuda_ipc_check_rcache` expects `uct_cuda_ipc_key_t *`

Reply:
can we change the key parameter in uct_cuda_ipc_check_rcache() to be const ?

Reply:
@yosefe I can make that change but there maybe one issue. ipc_check_rcache() calls rcache_get which may end up calling rcache_ops.mem_reg which expects void *arg but we would have to pass const void *arg. That said, mem_reg doesn't modify arg. 
```C
uct_cuda_ipc_rcache_mem_reg(void *context, ucs_rcache_t *rcache, void *arg, 
                            ucs_rcache_region_t *region, uint16_t flags);
```